lib/AST/ProtocolConformance.cpp

```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
   TypeWitnesses[assocType] = std::make_pair(substitution, typeDecl);
 }
 
+#ifndef NDEBUG
+static bool isSelfParameter(CanType type) {
```

Comment:
This will cause crashes on invalid code. Instead, always compare against proto->getSelfInterfaceType().